extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/SlicerUtils.java

```diff
@@ -98,6 +98,66 @@ public static <T> List<List<T>> makeSlices(
     return slicesList;
   }
 
+  /**
+   * Creates up to "maxNumSlices" lists from "iterator".
+   *
+   * This method creates as few slices as possible, while keeping each slice under the provided limits.
+   *
+   * If there is a conflict between maxNumSlices and maxFilesPerSlice or maxWeightPerSlice, then maxNumSlices wins.
+   * This means that for small values of maxNumSlices, slices may have more than maxFilesPerSlice files, or more
+   * than maxWeightPerSlice weight.
+   */
+  public static <T> List<List<T>> makeSlicesDynamic(
+      final Iterator<T> iterator,
+      final ToLongFunction<T> weightFunction,
+      final int maxNumSlices,
+      final int maxFilesPerSlice,
+      final long maxWeightPerSlice
+  )
+  {
```

Comment:
The algorithm here doesn't seem to be quite what one would expect.

We have a priority queue of the split lists. Cool. We fill the first split until it is full, then we start on the next one. It is not clear how if we have n splits with a max of m splits, we end up with n splits if n < m, and n splits of n/m blocks otherwise.

Also, we have a limit on the maximum number of slices. This means that, when the actual data exceeds that number, we need to make slices larger. (Or fail the ingest, or discard splits, or ...)

I guess I would have expected to see a loop that first sums the file sizes to get t. Then, divide that by the target split size to get the ideal number of splits. If that is too large, compute the actual number of splits by dividing the total by the maximum split count.

But, of course, we're ignoring the file count. We can't have any fewer splits than the number of files. So, a first check is to count files. For each, compute the ideal split count. This gives the candidate split count. Then, adjust if the ideal count is too high.

The result will be a set of files, with size and with splits. Now we can do bin packing for tasks based on these values. If the (adjusted) split count is less than the task limit, run one split per task. Else, evenly allocate sizes (not splits) per task so that each processes roughly total size / task count bytes.

It is not entirely clear how the algorithm here approximates what it seems we need. Maybe a few more comments might fill in the gaps?

Reply:
I'll add these comments to hopefully clarify things:

```
   * This function uses a greedy algorithm that starts out with a single empty slice. Items are assigned to the slice
   * that has the lowest weight at the time that item is encountered. New slices are created, up to maxNumSlices, when
   * the lightest existing slice exceeds either maxFilesPerSlice or maxWeightPerSlice.
   *
   * Slices may have more than maxFilesPerSlice files, or more than maxWeightPerSlice weight, if necessary to
   * keep the total number of slices at or below maxNumSlices.
```